### DISCUSSION TITLE
Plotly chart in reactpy app

### DISCUSSION THREAD

**Discussion by fsmosca** [idom-team/idom#1129] [Ideas]
I have this approach to create a plotly chart based from csv file input. Creates figure based from dataframe. Convert figure to html using plotly and finally generates vdom using reactpy.

Is this a fine approach? What are the other methods?

This is for bar chart.

```python
"""Renders plotly chart in ReactPy"""


from io import StringIO
from reactpy import component, html, utils
from reactpy.backend.fastapi import configure, Options
from fastapi import FastAPI
import plotly.express as px
import pandas as pd


DATA_FILENAME = 'expenses.csv'

PAGE_TITLE = 'ReactPy-PlotlyChart'

PLOTLY_JS = {
    'src': 'https://cdn.plot.ly/plotly-latest.min.js',
    'charset': 'utf-8'
}


@component
def BarChart():
    """Creates a plotly bar chart."""
    # Read csv file and convert to dataframe.
    df = pd.read_csv(DATA_FILENAME)

    # Group similar category and sum it.
    df_grp = df.groupby(['Category'])['Amount'].sum().reset_index()

    # Create a figure using the grouped dataframe.
    fig = px.bar(
        df_grp, x='Amount', y='Category', orientation='h',
        width=600, height=400, text_auto=True, title='Bar Chart'
    )

    # Create an html object in memory from fig.
    buffer = StringIO()
    fig.write_html(buffer, include_plotlyjs='cdn')
    fig_html = buffer.getvalue()

    return html.div(
        html.h2('Plotly Chart in ReactPy'),
        utils.html_to_vdom(fig_html)
    )


app = FastAPI()
configure(
    app,
    BarChart,
    Options(
        head=html.head(
            html.script(PLOTLY_JS),
            html.title(PAGE_TITLE)
        )
    )
)
```

You may download the [expenses.csv](https://drive.google.com/file/d/1z3yfIhaougnrln4dcw1Q_-U9wvB-sFhR/view?usp=drive_link) file from my google drive.

### Output

![](https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEgWiutt5QqQ0PQE7hQ6hY5DUzpzgs2_6d7l2FLBDPwM2MosHqADpCQC2HQEYSFBdcre4PJ194-8nuB4XgtptMpjnIZhzBqUTqtMkWgDVg5go9yE6N3vHxQsGML4j4qD5d4g3N9jPhYtpKy6pBZ5Hl0G10t1WT0IiCUkw0ky_WLac36NqhShbvLskaqScNRT/s1600/install-reactpy49.png)

I have created [plotly charts article](https://energybeam.blogspot.com/2023/08/how-to-add-plotly-charts-in-reactpy.html) about this.

**Comment by pharouhk** [comment]
Hi @fsmosca thanks for this. I wanted to know if plotly_js must be added to the head

**Comment by fsmosca** [comment]
Yeah it is in the head.

```
app = FastAPI()
configure(
    app,
    BarChart,
    Options(
        head=html.head(
            html.script(PLOTLY_JS),  # <======
            html.title(PAGE_TITLE)
        )
    )
)
```
